src/Integrations/Integrations/Logs/LogsIntegration.php

```diff
@@ -0,0 +1,256 @@
+use DDTrace\SpanData;
+use DDTrace\Util\ObjectKVStore;
+
+use function DDTrace\logs_correlation_trace_id;
+
+class LogsIntegration extends Integration
+{
+    const NAME = 'logs';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
+    {
+        return self::NAME;
+    }
+
+    public static function laminasLogLevelToString(int $logLevel, string $fallback): string
+    {
+        switch ($logLevel) {
+            case 0:
+                return 'emergency';
+            case 1:
+                return 'alert';
+            case 2:
+                return 'critical';
+            case 3:
+                return 'error';
+            case 4:
+                return 'warning';
+            case 5:
+                return 'notice';
+            case 6:
+                return 'info';
+            case 7:
+                return 'debug';
+            default:
+                return $fallback;
+        }
+    }
+
+    public static function getPlaceholders(
+        string $traceIdSubstitute = null,
+        string $spanIdSubstitute = null
+    ): array {
+        $placeholders = [
+            '%dd.trace_id%' => 'dd.trace_id="' . ($traceIdSubstitute ?? logs_correlation_trace_id()) . '"',
+            '%dd.span_id%'  => 'dd.span_id="' . ($spanIdSubstitute ?? dd_trace_peek_span_id()) . '"',
+        ];
+
+        $appName = ddtrace_config_app_name();
+        if ($appName) {
+            $placeholders['%dd.service%'] = 'dd.service="' . $appName . '"';
+        } else {
+            $placeholders['%dd.service%'] = '';
+        }
+
+        $currentContext = \DDTrace\current_context();
+        if ($currentContext['version']) {
+            $placeholders['%dd.version%'] = 'dd.version="' . $currentContext['version'] . '"';
+        } else {
+            $placeholders['%dd.version%'] = '';
+        }
+
+        if ($currentContext['env']) {
+            $placeholders['%dd.env%'] = 'dd.env="' . $currentContext['env'] . '"';
+        } else {
+            $placeholders['%dd.env%'] = '';
+        }
+
+        return $placeholders;
+    }
+
+    public static function messageContainsPlaceholders(string $message): bool
+    {
+        $placeholders = LogsIntegration::getPlaceholders();
+
+        foreach ($placeholders as $placeholder => $value) {
+            if (str_contains($message, $placeholder)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public static function appendTraceIdentifiersToMessage(
+        string $message,
+        string $traceIdSubstitute = null,
+        string $spanIdSubstitute = null
+    ): string {
+        $placeholders = LogsIntegration::getPlaceholders($traceIdSubstitute, $spanIdSubstitute);
+        LogsIntegration::replacePlaceholders($message, $placeholders);
+
+        if (strpos($message, 'dd.trace_id=') !== false) {
+            // Trace identifiers were already appended sometime in the call stack
+            return $message;
+        }
```

Comment:
I think it is necessary to do a check of some sort. For instance, when calling `Log::debug('...');` in Laravel with Monolog, the following hooks would be triggered (in order):
- `Illuminate\Log\LogManager::debug`
- `Illuminate\Log\Logger::debug`
- `Monolog\Logger::debug`

(they all implements `Psr\Log\LoggerInterface::debug`)

If we don't do any check at all, then the trace correlation identifiers would appended three times in this case. Alternatively **(this is the commit I just made)**, we could check for each trace correlation identifiers (trace id, span id, env, ...) whether it is already in the message, **but** this would certainly require custom parsing from the customer's end.